webofneeds/won-core/src/main/java/won/protocol/util/WonRdfUtils.java

```diff
@@ -279,30 +280,92 @@ public static String getTextMessage(Model model){
       return null;
     }
 
-    /**
-     * Returns the first won:hasTextMessage object, or null if none is found.
-     * @param wonMessage
-     * @return
-     */
-    public static String getTextMessage(final WonMessage wonMessage){
-      Dataset dataset = wonMessage.getCompleteDataset();
-      RDFNode node = getTextMessageForResource(dataset, wonMessage.getMessageURI());
-      if (node != null) return node.asLiteral().toString();
-      node = getTextMessageForResource(dataset, wonMessage.getCorrespondingRemoteMessageURI());
-      if (node != null) return node.asLiteral().toString();
-      URI forwardedMessageURI = wonMessage.getForwardedMessageURI();
-      if (forwardedMessageURI != null){
-        node = getTextMessageForResource(dataset, forwardedMessageURI);
-        if (node != null) return node.asLiteral().toString();
-        //get the remote message for the forwarded message
-        RDFNode msgNode = RdfUtils.findOnePropertyFromResource(dataset, forwardedMessageURI, WONMSG.HAS_CORRESPONDING_REMOTE_MESSAGE);
-        if (msgNode != null && msgNode.isResource()){
-          node = getTextMessageForResource(dataset, msgNode.asResource());
-          if (node != null) return node.asLiteral().toString();
-        }
+
+      /**
+       * Returns the first won:hasTextMessage object, or null if none is found.
+       * tries the message, its corresponding remote message, and any forwarded message,
+       * if any of those are contained in the dataset
+       *
+       * @param wonMessage
+       * @return
+       */
+      public static String getTextMessage(final WonMessage wonMessage) {
+          URI messageURI = wonMessage.getMessageURI();
+
+          //find the text message in the message, the remote message, or any forwarded message
+          String queryString =
+                  "prefix msg: <http://purl.org/webofneeds/message#>\n" +
+                          "prefix won: <http://purl.org/webofneeds/model#>\n" +
+                          "\n" +
+                          "SELECT distinct ?txt WHERE {\n" +
+                          "  {\n" +
+                          "    graph ?gA { ?msg won:hasTextMessage ?txt }\n" +
+                          "  } union {\n" +
+                          "    graph ?gB { ?msg2 msg:hasCorrespondingRemoteMessage ?msg }\n" +
+                          "    graph ?gA { ?msg won:hasTextMessage ?txt }\n" +
```

Comment:
doesnt this case not already include the first union with just: graph ?gA { ?msg won:hasTextMessage ?txt } ??

Reply:
you're right. messed up the msg\d variables